### DISCUSSION TITLE
Compilation errors when using newer C++ standards

### DISCUSSION THREAD

**Discussion by radioactiveman** [audacious-media-player/audacious#99] [Q&A] [answered]
Out of curiosity I have tried to compile Audacious with the gnu++20 standard.
Both audacious and audacious-plugins then fail to compile null checks like this:

```c++
String foo;
if (foo != nullptr)
...
```

Error message:
```
../src/libaudgui/infowin.cc:519:32: error: use of overloaded operator '!=' is ambiguous (with operand types 'String' and 'std::nullptr_t')
    g_return_if_fail (filename != nullptr);
                      ~~~~~~~~ ^  ~~~~~~~
/usr/include/glib-2.0/glib/gmessages.h:640:19: note: expanded from macro 'g_return_if_fail'
    if (G_LIKELY (expr)) \
                  ^~~~
/usr/include/glib-2.0/glib/gmacros.h:1243:59: note: expanded from macro 'G_LIKELY'
#define G_LIKELY(expr) (__builtin_expect (_G_BOOLEAN_EXPR(expr), 1))
                                                          ^~~~
/usr/include/glib-2.0/glib/gmacros.h:1242:66: note: expanded from macro '_G_BOOLEAN_EXPR'
#define _G_BOOLEAN_EXPR(expr) _G_BOOLEAN_EXPR_IMPL (__COUNTER__, expr)
                                                                 ^~~~
/usr/include/glib-2.0/glib/gmacros.h:1236:8: note: expanded from macro '_G_BOOLEAN_EXPR_IMPL'
   if (expr)                                    \
       ^~~~
../src/libaudcore/objects.h:155:10: note: candidate function
    bool operator==(const String & b) const { return raw_equal(raw, b.raw); }
         ^
../src/libaudgui/infowin.cc:519:32: note: built-in candidate operator!=(const char *, const char *)
    g_return_if_fail (filename != nullptr);
                               ^
../src/libaudgui/infowin.cc:519:32: note: built-in candidate operator!=(const volatile char *, const volatile char *)
```

How can this be solved? Ideally in our String class, so we do not have to change all explicit null checks to this which compiles fine.

```c++
String foo;
if (foo)
...
```

Another question I have is why do we use the gnu c++ standard.
Audacious and -plugins compile successfully with `c++17` for example, at least on Linux.

**Comment by jlindgren90** [comment]
I don't think we should be comparing a `String` with `nullptr`. I'd rather replace those occurrences with the boolean-style check.

Regarding the `gnu++` vs `c++` thing, I think there were some older versions of gcc that didn't fully support the `c++` variants. I tried gcc 5.4.0 just now and it worked fine though, so maybe it's not an issue any more.

**Comment by radioactiveman** [comment, marked as answer]
All right, thanks John. Fixed with https://github.com/audacious-media-player/audacious/commit/c67afc5e0385c6c89fef3eed9cca5670c0ff1fff and https://github.com/audacious-media-player/audacious-plugins/commit/6009062ac259428d682ccf79e3e2f95310b2138d.

Using c++20 instead of gnu++20 caused the Windows build to fail with the error `M_PI was not declared in this scope`.
So let's better keep using the gnu std. https://github.com/audacious-media-player/audacious/blob/b7f0019b99b2cd745721d1ff92bf7d5b32f43bb4/src/libaudcore/equalizer.cc#L64
